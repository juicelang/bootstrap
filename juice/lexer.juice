export 'position := {
	line: int
	column: int
}

export 'location := {
	start: position
	end: position
}

export 'token := {
	eof_token(location: location)

}
