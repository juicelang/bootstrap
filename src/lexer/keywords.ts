export const keywords = [
	"type",
	"fn",
	"if",
	"else",
	"import",
	"export",
	"class",
	"impl",
	"match",
	"as",
	"return",
	"for",
	"loop",
	"of",
	"break",
	"static",
	"foreign",
];
