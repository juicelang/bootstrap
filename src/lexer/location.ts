export interface location {
  line: number;
  column: number;
}
