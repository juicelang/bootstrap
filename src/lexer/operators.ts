export const OPERATOR_REGEX = /[:=,\.@#!?&|+\-*/%<>\^]/;

export const operators = [
  ":=",
  "=",
  ":",
  ",",
  ".",
  "@",
  "#",
  "!",
  "?",
  "&",
  "|",
  "&&",
  "||",
  "->",
  "+",
  "-",
  "*",
  "/",
  "%",
  "**",
  "++",
  "--",
  ">=",
  "<=",
  "==",
  "!=",
  ">",
  "<",
  "^",
] as const;

export const precedence: Record<(typeof operators)[number], number> = {
  ":=": 0,
  "=": 0,
  ":": 0,
  ",": 0,
  ".": 0,
  "@": 0,
  "#": 0,
  "!": 0,
  "?": 0,
  "&": 0,
  "|": 0,
  "&&": 0,
  "||": 0,
  "->": 0,
  "+": 0,
  "-": 0,
  "*": 0,
  "/": 0,
  "%": 0,
  "**": 0,
  "++": 0,
  "--": 0,
  ">=": 0,
  "<=": 0,
  "==": 0,
  "!=": 0,
  ">": 0,
  "<": 0,
  "^": 0,
};

export function is_operator(text: string): text is (typeof operators)[number] {
  return operators.includes(text as any);
}
