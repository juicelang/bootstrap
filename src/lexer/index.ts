import { keywords } from "./keywords";
import { location } from "./location";
import { OPERATOR_REGEX, is_operator } from "./operators";
import * as tokens from "./tokens";

export default class lexer {
  cursor = 0;
  line = 1;
  column = 1;
  code = "";

  location(): location {
    return {
      line: this.line,
      column: this.column,
    };
  }

  eat() {
    let c = this.code[this.cursor];

    this.cursor++;

    if (c === "\n") {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }

    return c;
  }

  peek(offset = 0) {
    let c = this.code[this.cursor + offset];

    return c;
  }

  whitespace() {
    let ws = "";

    while (
      this.peek() === " " ||
      this.peek() === "\t" ||
      this.peek() === "\n"
    ) {
      ws += this.eat();
    }

    return ws;
  }

  lex(
    code: string,
    {
      cursor = 0,
      line = 1,
      column = 1,
    }: {
      cursor?: number;
      line?: number;
      column?: number;
    } = {},
  ): tokens.token[] {
    this.cursor = cursor;
    this.line = line;
    this.column = column;
    this.code = code;

    let ts: tokens.token[] = [];

    while (this.cursor < this.code.length) {
      let t = this.lex_token();

      ts.push(t);

      if (t.kind === "eof") {
        break;
      }
    }

    let last_token = ts[ts.length - 1];

    if (!last_token || last_token.kind !== "eof") {
      ts.push({
        kind: "eof",
        location: { start: this.location(), end: this.location() },
      });
    }

    return ts;
  }

  lex_token(): tokens.token {
    let c = this.peek();
    let next = this.peek(1);

    if (c === "{") {
      return this.lex_open_curly();
    } else if (c === "}") {
      return this.lex_close_curly();
    } else if (c === "(") {
      return this.lex_open_paren();
    } else if (c === ")") {
      return this.lex_close_paren();
    } else if (c === "[") {
      return this.lex_open_bracket();
    } else if (c === "]") {
      return this.lex_close_bracket();
    } else if (c === "/" && next && next === "/") {
      return this.lex_comment();
    } else if (OPERATOR_REGEX.test(c)) {
      return this.lex_operator();
    } else if (c.match(/\s/)) {
      return this.lex_whitespace();
    } else if (c === "$" && next === "{") {
      return this.lex_interpolation();
    } else if (c === "$") {
      return this.lex_type_identifier();
    } else if (c.match(/[a-zA-Z_]/)) {
      return this.lex_identifier();
    } else if (c.match(/[0-9]/)) {
      return this.lex_number();
    } else if (c === '"') {
      return this.lex_string();
    } else {
      this.eat();

      throw new Error(`no token found for "${c}"`);
    }
  }

  lex_whitespace(): tokens.whitespace_token {
    let start = this.location();

    let value = this.whitespace();

    let end = this.location();

    return {
      kind: "whitespace",
      location: { start, end },
      value,
    };
  }

  lex_open_curly(): tokens.open_curly_token {
    let start = this.location();
    let end = this.location();

    this.eat();

    return {
      kind: "open_curly",
      location: { start, end },
    };
  }

  lex_close_curly(): tokens.close_curly_token {
    let start = this.location();
    let end = this.location();

    this.eat();

    return {
      kind: "close_curly",
      location: { start, end },
    };
  }

  lex_open_paren(): tokens.open_paren_token {
    let start = this.location();
    let end = this.location();

    this.eat();

    return {
      kind: "open_paren",
      location: { start, end },
    };
  }

  lex_close_paren(): tokens.close_paren_token {
    let start = this.location();
    let end = this.location();

    this.eat();

    return {
      kind: "close_paren",
      location: { start, end },
    };
  }

  lex_open_bracket(): tokens.open_bracket_token {
    let start = this.location();
    let end = this.location();

    this.eat();

    return {
      kind: "open_bracket",
      location: { start, end },
    };
  }

  lex_close_bracket(): tokens.close_bracket_token {
    let start = this.location();
    let end = this.location();

    this.eat();

    return {
      kind: "close_bracket",
      location: { start, end },
    };
  }

  lex_comment(): tokens.comment_token {
    let start = this.location();

    this.eat();
    this.eat();

    let value = "";

    while (this.peek() && this.peek() !== "\n") {
      value += this.eat();
    }

    let end = this.location();

    return {
      kind: "comment",
      location: { start, end },
      value,
    };
  }

  lex_identifier():
    | tokens.identifier_token
    | tokens.macro_identifier_token
    | tokens.keyword_token {
    let start = this.location();
    let value = this.eat();
    let kind:
      | tokens.identifier_token["kind"]
      | tokens.macro_identifier_token["kind"]
      | tokens.keyword_token["kind"] = "identifier";

    while (this.peek() && this.peek().match(/[a-zA-Z0-9_]/)) {
      value += this.eat();
    }

    let next = this.peek();

    if (next === "!") {
      kind = "macro_identifier";
      value += this.eat();
    } else if (next === "'") {
      value += this.eat();
    } else if (keywords.includes(value)) {
      kind = "keyword";
    }

    let end = this.location();

    return {
      kind,
      location: { start, end },
      value,
    };
  }

  lex_type_identifier(): tokens.type_identifier_token {
    let start = this.location();
    let value = this.eat();

    while (this.peek() && this.peek().match(/[a-zA-Z0-9_]/)) {
      value += this.eat();
    }

    let end = this.location();

    return {
      kind: "type_identifier",
      location: { start, end },
      value,
    };
  }

  lex_operator(): tokens.operator_token {
    let start = this.location();
    let value = this.eat();

    while (this.peek() && OPERATOR_REGEX.test(this.peek())) {
      value += this.eat();
    }

    let end = this.location();

    if (!is_operator(value)) {
      throw new Error(`invalid operator "${value}"`);
    }

    return {
      kind: "operator",
      location: { start, end },
      value,
    };
  }

  lex_number(): tokens.number_token {
    let start = this.location();

    let raw = "";
    let value = "";
    let variant: tokens.number_token["variant"] = "integer";

    let number_regex = /[0-9\._]/;

    let c = this.peek();
    if (c === "0") {
      let next = this.peek(1);

      if (next === "x") {
        number_regex = /[0-9a-fA-F_]/;
        variant = "hex";
        raw += this.eat();
        raw += this.eat();
      } else if (next === "o") {
        number_regex = /[0-7_]/;
        variant = "octal";
        raw += this.eat();
        raw += this.eat();
      } else if (next === "b") {
        number_regex = /[0-1_]/;
        variant = "binary";
        raw += this.eat();
        raw += this.eat();
      }
    }

    while (this.peek() && this.peek().match(number_regex)) {
      let c = this.peek();

      let next = this.peek(1);

      if (
        c === "_" &&
        value[value.length - 1] !== "_" &&
        next &&
        next.match(number_regex) &&
        next !== "_"
      ) {
        raw += this.eat();
        continue;
      }

      // If we aren't parsing a float or we are parsing something like a hex, octal, or
      // binary number, then this is a property accessor.
      if (c === "." && (variant === "float" || variant !== "integer")) {
        break;
      }

      if (c === "." && next && !next.match(number_regex)) {
        break;
      }

      if (c === ".") {
        variant = "float";
      }

      raw += c;
      value += this.eat();
    }

    let end = this.location();

    let number: number;

    switch (variant) {
      default:
      case "integer":
        number = parseInt(value, 10);
        break;
      case "float":
        number = parseFloat(value);
        break;
      case "hex":
        number = parseInt(value, 16);
        break;
      case "octal":
        number = parseInt(value, 8);
        break;
      case "binary":
        number = parseInt(value, 2);
        break;
    }

    return {
      kind: "number",
      location: { start, end },
      value: number,
      variant,
      raw,
    };
  }

  lex_interpolation(): tokens.interpolation_token {
    this.eat();
    this.eat();

    let interpolated_start = this.location();
    let interpolated_end: location | undefined;

    let ts: tokens.token[] = [];

    let depth = 1;

    while (depth > 0 && this.peek()) {
      let next = this.peek();

      if (next === "{") {
        depth++;
      } else if (next === "}") {
        depth--;
      }

      if (depth > 0) {
        ts.push(this.lex_token());
      } else {
        interpolated_end = this.location();
        this.eat();
      }
    }

    if (interpolated_end === undefined) {
      interpolated_end = this.location();
    }

    return {
      kind: "interpolation",
      location: { start: interpolated_start, end: interpolated_end },
      value: ts,
    };
  }

  lex_string(): tokens.string_token {
    let start = this.location();

    this.eat();

    let value: tokens.string_token["value"] = [];

    while (this.peek() && this.peek() !== '"') {
      let c = this.peek();

      if (c === "$" && this.peek(1) === "{") {
        value.push(this.lex_interpolation());

        value.push({
          kind: "raw_string",
          location: { start: this.location(), end: this.location() },
          value: "",
        });
      } else {
        let last_token = value[value.length - 1];

        let start = this.location();

        let c = this.eat();

        if (c === "\\") {
          c = this.eat();
        }

        if (last_token && last_token.kind === "raw_string") {
          last_token.value += c;
          last_token.location.end = this.location();
        } else {
          value.push({
            kind: "raw_string",
            location: { start, end: this.location() },
            value: c,
          });
        }
      }
    }

    this.eat();

    let end = this.location();

    return {
      kind: "string",
      location: { start, end },
      value,
    };
  }
}
