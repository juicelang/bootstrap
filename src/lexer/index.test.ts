import { describe, test, expect, beforeEach } from "bun:test";
import lexer from "./index";
import { operators } from "./operators";
import * as tokens from "./tokens";

const trim = (s: string) => s.trim();

describe("Lexer", () => {
  let l: lexer;

  beforeEach(() => {
    l = new lexer();
  });

  describe("Tokens", () => {
    describe("Identifiers", () => {
      test("lexes plain identifiers", () => {
        let code = trim(`x`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "identifier",
            value: "x",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 2 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 2 },
              end: { line: 1, column: 2 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes long identifiers", () => {
        let code = trim(`this_is_an_identifier`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "identifier",
            value: "this_is_an_identifier",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 22 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 22 },
              end: { line: 1, column: 22 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes prime identifiers", () => {
        let code = trim(`x'`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "identifier",
            value: "x'",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 3 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 3 },
              end: { line: 1, column: 3 },
            },
          },
        ] satisfies tokens.token[]);
      });
    });

    describe("Primitives", () => {
      test("lexes integers", () => {
        let code = trim(`42`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "number",
            value: 42,
            variant: "integer",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 3 },
            },
            raw: "42",
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 3 },
              end: { line: 1, column: 3 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes floats", () => {
        let code = trim(`42.0`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "number",
            value: 42.0,
            variant: "float",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 5 },
            },
            raw: "42.0",
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 5 },
              end: { line: 1, column: 5 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes hexadecimals", () => {
        let code = trim(`0x1`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "number",
            value: 1,
            variant: "hex",
            raw: "0x1",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 4 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 4 },
              end: { line: 1, column: 4 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes octals", () => {
        let code = trim(`0o1`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "number",
            value: 1,
            variant: "octal",
            raw: "0o1",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 4 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 4 },
              end: { line: 1, column: 4 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes binary", () => {
        let code = trim(`0b1`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "number",
            value: 1,
            variant: "binary",
            raw: "0b1",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 4 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 4 },
              end: { line: 1, column: 4 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes long numbers", () => {
        let code = trim(`123_456_789.0_1_2_3`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "number",
            value: 123456789.0123,
            variant: "float",
            raw: "123_456_789.0_1_2_3",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 20 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 20 },
              end: { line: 1, column: 20 },
            },
          },
        ] satisfies tokens.token[]);
      });
    });

    describe("Comments", () => {
      test("lexes comments", () => {
        let code = trim(`// hello world`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "comment",
            value: " hello world",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 15 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 15 },
              end: { line: 1, column: 15 },
            },
          },
        ] satisfies tokens.token[]);
      });
    });

    describe("Strings", () => {
      test("lexes strings", () => {
        let code = trim(`"hello world"`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "string",
            value: [
              {
                kind: "raw_string",
                value: "hello world",
                location: {
                  start: { line: 1, column: 2 },
                  end: { line: 1, column: 13 },
                },
              },
            ],
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 14 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 14 },
              end: { line: 1, column: 14 },
            },
          },
        ] satisfies tokens.token[]);
      });

      test("lexes interpolations", () => {
        let code = trim(`"hello \${name}"`);
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "string",
            value: [
              {
                kind: "raw_string",
                value: "hello ",
                location: {
                  start: { line: 1, column: 2 },
                  end: { line: 1, column: 8 },
                },
              },
              {
                kind: "interpolated_string",
                value: [
                  {
                    kind: "identifier",
                    value: "name",
                    location: {
                      start: { line: 1, column: 10 },
                      end: { line: 1, column: 14 },
                    },
                  },
                ],
                location: {
                  start: { line: 1, column: 10 },
                  end: { line: 1, column: 14 },
                },
              },
            ],
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 16 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 16 },
              end: { line: 1, column: 16 },
            },
          },
        ] satisfies tokens.token[]);
      });
    });

    describe("Keywords", () => {
      test("lexes fn", () => {
        let code = trim("fn");
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "keyword",
            value: "fn",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 3 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 3 },
              end: { line: 1, column: 3 },
            },
          },
        ]);
      });
    });

    describe("Operators", () => {
      test("lexes operators", () => {
        let code = trim(operators.join(" "));
        let tokens = l.lex(code);

        expect(tokens).toEqual([
          {
            kind: "operator",
            value: ":=",
            location: {
              start: { line: 1, column: 1 },
              end: { line: 1, column: 3 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 3 },
              end: { line: 1, column: 4 },
            },
          },
          {
            kind: "operator",
            value: "=",
            location: {
              start: { line: 1, column: 4 },
              end: { line: 1, column: 5 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 5 },
              end: { line: 1, column: 6 },
            },
          },
          {
            kind: "operator",
            value: ":",
            location: {
              start: { line: 1, column: 6 },
              end: { line: 1, column: 7 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 7 },
              end: { line: 1, column: 8 },
            },
          },
          {
            kind: "operator",
            value: ",",
            location: {
              start: { line: 1, column: 8 },
              end: { line: 1, column: 9 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 9 },
              end: { line: 1, column: 10 },
            },
          },
          {
            kind: "operator",
            value: ".",
            location: {
              start: { line: 1, column: 10 },
              end: { line: 1, column: 11 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 11 },
              end: { line: 1, column: 12 },
            },
          },
          {
            kind: "operator",
            value: "@",
            location: {
              start: { line: 1, column: 12 },
              end: { line: 1, column: 13 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 13 },
              end: { line: 1, column: 14 },
            },
          },
          {
            kind: "operator",
            value: "#",
            location: {
              start: { line: 1, column: 14 },
              end: { line: 1, column: 15 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 15 },
              end: { line: 1, column: 16 },
            },
          },
          {
            kind: "operator",
            value: "!",
            location: {
              start: { line: 1, column: 16 },
              end: { line: 1, column: 17 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 17 },
              end: { line: 1, column: 18 },
            },
          },
          {
            kind: "operator",
            value: "?",
            location: {
              start: { line: 1, column: 18 },
              end: { line: 1, column: 19 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 19 },
              end: { line: 1, column: 20 },
            },
          },
          {
            kind: "operator",
            value: "&",
            location: {
              start: { line: 1, column: 20 },
              end: { line: 1, column: 21 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 21 },
              end: { line: 1, column: 22 },
            },
          },
          {
            kind: "operator",
            value: "|",
            location: {
              start: { line: 1, column: 22 },
              end: { line: 1, column: 23 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 23 },
              end: { line: 1, column: 24 },
            },
          },
          {
            kind: "operator",
            value: "&&",
            location: {
              start: { line: 1, column: 24 },
              end: { line: 1, column: 26 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 26 },
              end: { line: 1, column: 27 },
            },
          },
          {
            kind: "operator",
            value: "||",
            location: {
              start: { line: 1, column: 27 },
              end: { line: 1, column: 29 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 29 },
              end: { line: 1, column: 30 },
            },
          },
          {
            kind: "operator",
            value: "->",
            location: {
              start: { line: 1, column: 30 },
              end: { line: 1, column: 32 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 32 },
              end: { line: 1, column: 33 },
            },
          },
          {
            kind: "operator",
            value: "+",
            location: {
              start: { line: 1, column: 33 },
              end: { line: 1, column: 34 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 34 },
              end: { line: 1, column: 35 },
            },
          },
          {
            kind: "operator",
            value: "-",
            location: {
              start: { line: 1, column: 35 },
              end: { line: 1, column: 36 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 36 },
              end: { line: 1, column: 37 },
            },
          },
          {
            kind: "operator",
            value: "*",
            location: {
              start: { line: 1, column: 37 },
              end: { line: 1, column: 38 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 38 },
              end: { line: 1, column: 39 },
            },
          },
          {
            kind: "operator",
            value: "/",
            location: {
              start: { line: 1, column: 39 },
              end: { line: 1, column: 40 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 40 },
              end: { line: 1, column: 41 },
            },
          },
          {
            kind: "operator",
            value: "%",
            location: {
              start: { line: 1, column: 41 },
              end: { line: 1, column: 42 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 42 },
              end: { line: 1, column: 43 },
            },
          },
          {
            kind: "operator",
            value: "**",
            location: {
              start: { line: 1, column: 43 },
              end: { line: 1, column: 45 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 45 },
              end: { line: 1, column: 46 },
            },
          },
          {
            kind: "operator",
            value: "++",
            location: {
              start: { line: 1, column: 46 },
              end: { line: 1, column: 48 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 48 },
              end: { line: 1, column: 49 },
            },
          },
          {
            kind: "operator",
            value: "--",
            location: {
              start: { line: 1, column: 49 },
              end: { line: 1, column: 51 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 51 },
              end: { line: 1, column: 52 },
            },
          },
          {
            kind: "operator",
            value: ">=",
            location: {
              start: { line: 1, column: 52 },
              end: { line: 1, column: 54 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 54 },
              end: { line: 1, column: 55 },
            },
          },
          {
            kind: "operator",
            value: "<=",
            location: {
              start: { line: 1, column: 55 },
              end: { line: 1, column: 57 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 57 },
              end: { line: 1, column: 58 },
            },
          },
          {
            kind: "operator",
            value: "==",
            location: {
              start: { line: 1, column: 58 },
              end: { line: 1, column: 60 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 60 },
              end: { line: 1, column: 61 },
            },
          },
          {
            kind: "operator",
            value: "!=",
            location: {
              start: { line: 1, column: 61 },
              end: { line: 1, column: 63 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 63 },
              end: { line: 1, column: 64 },
            },
          },
          {
            kind: "operator",
            value: ">",
            location: {
              start: { line: 1, column: 64 },
              end: { line: 1, column: 65 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 65 },
              end: { line: 1, column: 66 },
            },
          },
          {
            kind: "operator",
            value: "<",
            location: {
              start: { line: 1, column: 66 },
              end: { line: 1, column: 67 },
            },
          },
          {
            kind: "whitespace",
            value: " ",
            location: {
              start: { line: 1, column: 67 },
              end: { line: 1, column: 68 },
            },
          },
          {
            kind: "operator",
            value: "^",
            location: {
              start: { line: 1, column: 68 },
              end: { line: 1, column: 69 },
            },
          },
          {
            kind: "eof",
            location: {
              start: { line: 1, column: 69 },
              end: { line: 1, column: 69 },
            },
          },
        ] satisfies tokens.token[]);
      });
    });
  });

  describe("Examples", () => {
    test("lexes hello world", () => {
      let code = trim(`
				import dev.juice.io (print_line)

				export fn main() {
					print_line("hello world!")
				}
			`);
      let tokens = l.lex(code);

      expect(tokens).toMatchSnapshot();
    });
  });
});
