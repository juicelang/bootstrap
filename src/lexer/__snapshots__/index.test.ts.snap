// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Lexer Examples lexes hello world 1`] = `
[
  {
    "kind": "keyword",
    "location": {
      "end": {
        "column": 7,
        "line": 1,
      },
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "value": "import",
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 8,
        "line": 1,
      },
      "start": {
        "column": 7,
        "line": 1,
      },
    },
    "value": " ",
  },
  {
    "kind": "identifier",
    "location": {
      "end": {
        "column": 11,
        "line": 1,
      },
      "start": {
        "column": 8,
        "line": 1,
      },
    },
    "value": "dev",
  },
  {
    "kind": "operator",
    "location": {
      "end": {
        "column": 12,
        "line": 1,
      },
      "start": {
        "column": 11,
        "line": 1,
      },
    },
    "value": ".",
  },
  {
    "kind": "identifier",
    "location": {
      "end": {
        "column": 17,
        "line": 1,
      },
      "start": {
        "column": 12,
        "line": 1,
      },
    },
    "value": "juice",
  },
  {
    "kind": "operator",
    "location": {
      "end": {
        "column": 18,
        "line": 1,
      },
      "start": {
        "column": 17,
        "line": 1,
      },
    },
    "value": ".",
  },
  {
    "kind": "identifier",
    "location": {
      "end": {
        "column": 20,
        "line": 1,
      },
      "start": {
        "column": 18,
        "line": 1,
      },
    },
    "value": "io",
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 21,
        "line": 1,
      },
      "start": {
        "column": 20,
        "line": 1,
      },
    },
    "value": " ",
  },
  {
    "kind": "open_paren",
    "location": {
      "end": {
        "column": 21,
        "line": 1,
      },
      "start": {
        "column": 21,
        "line": 1,
      },
    },
  },
  {
    "kind": "identifier",
    "location": {
      "end": {
        "column": 32,
        "line": 1,
      },
      "start": {
        "column": 22,
        "line": 1,
      },
    },
    "value": "print_line",
  },
  {
    "kind": "close_paren",
    "location": {
      "end": {
        "column": 32,
        "line": 1,
      },
      "start": {
        "column": 32,
        "line": 1,
      },
    },
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 5,
        "line": 3,
      },
      "start": {
        "column": 33,
        "line": 1,
      },
    },
    "value": 
"

				"
,
  },
  {
    "kind": "keyword",
    "location": {
      "end": {
        "column": 11,
        "line": 3,
      },
      "start": {
        "column": 5,
        "line": 3,
      },
    },
    "value": "export",
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 12,
        "line": 3,
      },
      "start": {
        "column": 11,
        "line": 3,
      },
    },
    "value": " ",
  },
  {
    "kind": "keyword",
    "location": {
      "end": {
        "column": 14,
        "line": 3,
      },
      "start": {
        "column": 12,
        "line": 3,
      },
    },
    "value": "fn",
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 15,
        "line": 3,
      },
      "start": {
        "column": 14,
        "line": 3,
      },
    },
    "value": " ",
  },
  {
    "kind": "identifier",
    "location": {
      "end": {
        "column": 19,
        "line": 3,
      },
      "start": {
        "column": 15,
        "line": 3,
      },
    },
    "value": "main",
  },
  {
    "kind": "open_paren",
    "location": {
      "end": {
        "column": 19,
        "line": 3,
      },
      "start": {
        "column": 19,
        "line": 3,
      },
    },
  },
  {
    "kind": "close_paren",
    "location": {
      "end": {
        "column": 20,
        "line": 3,
      },
      "start": {
        "column": 20,
        "line": 3,
      },
    },
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 22,
        "line": 3,
      },
      "start": {
        "column": 21,
        "line": 3,
      },
    },
    "value": " ",
  },
  {
    "kind": "open_curly",
    "location": {
      "end": {
        "column": 22,
        "line": 3,
      },
      "start": {
        "column": 22,
        "line": 3,
      },
    },
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 6,
        "line": 4,
      },
      "start": {
        "column": 23,
        "line": 3,
      },
    },
    "value": 
"
					"
,
  },
  {
    "kind": "identifier",
    "location": {
      "end": {
        "column": 16,
        "line": 4,
      },
      "start": {
        "column": 6,
        "line": 4,
      },
    },
    "value": "print_line",
  },
  {
    "kind": "open_paren",
    "location": {
      "end": {
        "column": 16,
        "line": 4,
      },
      "start": {
        "column": 16,
        "line": 4,
      },
    },
  },
  {
    "kind": "string",
    "location": {
      "end": {
        "column": 31,
        "line": 4,
      },
      "start": {
        "column": 17,
        "line": 4,
      },
    },
    "value": [
      {
        "kind": "raw_string",
        "location": {
          "end": {
            "column": 30,
            "line": 4,
          },
          "start": {
            "column": 18,
            "line": 4,
          },
        },
        "value": "hello world!",
      },
    ],
  },
  {
    "kind": "close_paren",
    "location": {
      "end": {
        "column": 31,
        "line": 4,
      },
      "start": {
        "column": 31,
        "line": 4,
      },
    },
  },
  {
    "kind": "whitespace",
    "location": {
      "end": {
        "column": 5,
        "line": 5,
      },
      "start": {
        "column": 32,
        "line": 4,
      },
    },
    "value": 
"
				"
,
  },
  {
    "kind": "close_curly",
    "location": {
      "end": {
        "column": 5,
        "line": 5,
      },
      "start": {
        "column": 5,
        "line": 5,
      },
    },
  },
  {
    "kind": "eof",
    "location": {
      "end": {
        "column": 6,
        "line": 5,
      },
      "start": {
        "column": 6,
        "line": 5,
      },
    },
  },
]
`;
